staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go

```diff
@@ -225,11 +229,15 @@ func RunOrDie(ctx context.Context, lec LeaderElectionConfig) {
 // GetLeader returns the identity of the last observed leader or returns the empty string if
 // no leader has yet been observed.
 func (le *LeaderElector) GetLeader() string {
+	le.mutex.RLock()
+	defer le.mutex.RUnlock()
 	return le.observedRecord.HolderIdentity
```

Comment:
I'm not sure what the `GetLeader()` function is used for. The application that uses leader election should put its logic into those callback functions. No one should poll the `GetLeader()` function periodically and react on it. IMO we may deprecate this function and make `IsLeader` a private function.

Even if we really want this function (maybe for some monitoring that runs aside?), I'm not sure if the existing implementation is correct. The cached `observedRecord` could be stale if a leader lost the leadership but somehow got stuck and didn't return. One can use the resourcelock interface to retrieve the up-to-date leader's identity (though it queries apiserver so it's more expensive), and this function won't depend on `observedRecord`.

@mikedanese WDYT?

Reply:
Sounds good. @d-kuro could you update the comments to document that `GetLeader()` is for informational purposes?

Reply:
updated: https://github.com/kubernetes/kubernetes/pull/97419/commits/0b8181a0f2a219eb6c26ae984eb90bfa961f84e1

Do you have any concerns?